### ISSUE TITLE
Incompatible with early access releases of Java (Java >= 16 currently)

### ISSUE THREAD

**Issue by magneticflux-** [CaffeineMC/hydrogen-fabric#13]
As it turns out, the JVM does something a bit worse than a nasty warning:
```
[07:05:29] [main/WARN]: Trying to reflect-hack into the class-loader so we can define our own classes in an unrestricted loader...
[07:05:29] [main/WARN]: ** The JVM might warn you with something nasty! **
Failed to start Minecraft:
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.multimc.onesix.OneSixLauncher.launchWithMainClass(OneSixLauncher.java:196)
	at org.multimc.onesix.OneSixLauncher.launch(OneSixLauncher.java:231)
	at org.multimc.EntryPoint.listen(EntryPoint.java:143)
	at org.multimc.EntryPoint.main(EntryPoint.java:34)
Caused by: java.lang.RuntimeException: Could not execute entrypoint stage 'preLaunch' due to errors, provided by 'hydrogen'!
	at net.fabricmc.loader.entrypoint.minecraft.hooks.EntrypointUtils.invoke0(EntrypointUtils.java:53)
	at net.fabricmc.loader.entrypoint.minecraft.hooks.EntrypointUtils.invoke(EntrypointUtils.java:36)
	at net.fabricmc.loader.launch.knot.Knot.init(Knot.java:137)
	at net.fabricmc.loader.launch.knot.KnotClient.main(KnotClient.java:27)
	... 8 more
Caused by: java.lang.ExceptionInInitializerError
	at me.jellysquid.mods.hydrogen.common.jvm.ClassConstructors.initGuavaExtensions(ClassConstructors.java:24)
	at me.jellysquid.mods.hydrogen.common.jvm.ClassConstructors.init(ClassConstructors.java:15)
	at me.jellysquid.mods.hydrogen.common.HydrogenPreLaunch.onPreLaunch(HydrogenPreLaunch.java:12)
	at net.fabricmc.loader.entrypoint.minecraft.hooks.EntrypointUtils.invoke0(EntrypointUtils.java:50)
	... 11 more
Caused by: java.lang.RuntimeException: Could not expose method
	at me.jellysquid.mods.hydrogen.common.jvm.ClassDefineTool.<clinit>(ClassDefineTool.java:25)
	... 15 more
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(byte[],int,int) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @1b6924cb
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:357)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
	at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:199)
	at java.base/java.lang.reflect.Method.setAccessible(Method.java:193)
	at me.jellysquid.mods.hydrogen.common.jvm.ClassDefineTool.<clinit>(ClassDefineTool.java:21)
	... 15 more
```

This isn't likely fixable in code without fundamentally changing how the classes are defined/loaded, but it _is_ fixable with a certain combination of JVM flags for now.

This flag allows Hydrogen to access the protected method to define new classes:
```
--add-opens java.base/java.lang=ALL-UNNAMED
```

**Comment by magneticflux-** [comment]
This was introduced by JEP 396: https://openjdk.java.net/jeps/396

In it, they propose a workaround for uses of the `protected defineClass` method:

> Frameworks that use the protected defineClass methods of java.lang.ClassLoader in order to define new classes in existing class loaders. Such frameworks should instead use java.lang.invoke.MethodHandles.Lookup::defineClass, which has been available since JDK 9.

This may be done conditionally by checking the Java version at runtime, allowing for compatibility with Java 8 at the same time.

**Comment by jacklollz2** [comment]
Can confirm, this issue is persistent on JRE 16.

**Comment by Cognizaunce** [comment]
Did not want to make a separate issue, thought this might be relevant. Does the jvm implementation matter? If I use openj9 instead of hotspot, will the mod still behave as intended?

**Comment by jacklollz2** [comment]
> Did not want to make a separate issue, thought this might be relevant. Does the jvm implementation matter? If I use openj9 instead of hotspot, will the mod still behave as intended?

For Minecraft: The implementation does matter, official java resources are designed for Minecraft, not eclipses variant.

**Comment by jellysquid3** [comment]
This is now solved with the release for Minecraft 1.17 (by necessity.) I would recommend users upgrade to that version if they're wanting to use Java 16.
